Fix herder item stack request corruption

rivon0507 opened the pull request:
Closes #10884 

# Changes proposed in this pull request

- Prevent item stack corruption in herder AI by requesting single-item stacks alongside the count parameter instead of an oversized stack that get corrupted during network transmission. (Inspired by the postbox)

- Use placeholder values in error logging for deserialization issues.

## Testing
- [x] Yes I tested this before submitting it.
- [ ] I also did a multiplayer test.

Review please

rivon0507 committed:
Fix herder item request corruption

Prevent item stack corruption in herder AI by requesting single-item stacks
with separate count parameters instead of oversized stacks that get corrupted
during network transmission. (Inspired by the postbox)

Use placeholder values or variable substitution in error logging for deserialization issues.

Raycoms commented on src/main/java/com/minecolonies/core/entity/ai/workers/production/herders/AbstractEntityAIHerder.java:
```diff
@@ -314,7 +314,7 @@ private IAIState prepareForHerding()
         for (final ItemStack item : getExtraItemsNeeded())
         {
-            checkIfRequestForItemExistOrCreateAsync(item);
+            checkIfRequestForItemExistOrCreateAsync(item.copyWithCount(1), item.getCount(), item.getCount());
```
Ohhh, that's why it's running into the problem. I think we should probably
a) Inside the request creation do this, and log if the itemstack does exceed the 64 with stacktrace where this was created.
b) Make getExtraItemsNeeded an actual ItemStorage(ItemStack, qty). To avoid creating bad stacks to begin with here.

rivon0507 replied:
Does the item count in the `ItemStack` matter in a request, in the request system? If not (which is probably the case) the constructor of `com.minecolonies.api.colony.requestsystem.requestable.Stack` can be altered so that the `ItemStack` will always have a count of `1`, since that data is already available in `Stack.getCount()`.

And I'll make getExtraItemsNeeded return ItemStorage now.

Raycoms replied:
I'm not 100% sure if it matters. But overall we should never make too large stacks and log if we accidentially do from somewhere.

rivon0507 replied:
In `com.minecolonies.api.colony.requestsystem.requestable.Stack`, after looking up usages of `theStack` and ensuring no logic depends on `ItemStack.getCount()`, `theStack`'s count is set to `1` in the constructor of `Stack`. That way, we'll never have too large `ItemStack`s in requests.

And `getExtraItemsNeeded()` now returns an `ItemStorage`.

rivon0507 committed:
Prevent network corruption in Stack deliverables and refactor herder AI

- Migrate extra items needed from ItemStack to ItemStorage in herder AI
- Force ItemStack count to 1 in Stack deliverables to prevent network corruption
- Add logging for ItemStack.count and amount inconsistencies in Stack deliverables
